src/on_shutdown.c

```diff
@@ -0,0 +1,138 @@
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY AUTHORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "on_shutdown.h"
+
+#include <stdlib.h>
+#include <small/rlist.h>
+#include <errno.h>
+
+struct on_shutdown_trigger {
+	struct trigger trigger;
+	/** Shutdown trigger function */
+	void (*handler)(void *);
+	/** Trigger function argument */
+	void *arg;
+};
+
+/**
+ * Shutdown trigger list, not need mutex, because access is
+ * avaliable only from tx thread
+ */
+static struct rlist on_shutdown_trigger_list =
+	RLIST_HEAD_INITIALIZER(on_shutdown_trigger_list);
+
+static int
+trigger_commom_f(struct trigger *trigger, MAYBE_UNUSED void *event)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger =
+		container_of(trigger, struct on_shutdown_trigger, trigger);
+	on_shutdown_trigger->handler(on_shutdown_trigger->arg);
+	return 0;
+}
+
+static int
+on_shutdown_trigger_create(void (*handler)(void *), void *arg)
+{
+	struct on_shutdown_trigger *trigger = (struct on_shutdown_trigger *)
+		malloc(sizeof(struct on_shutdown_trigger));
+	if (trigger == NULL)
+		return -1;
+	trigger_create(&trigger->trigger, trigger_commom_f, NULL, NULL);
+	trigger->handler = handler;
+	trigger->arg = arg;
+	trigger_add(&on_shutdown_trigger_list, &trigger->trigger);
+	return 0;
+}
+
+/**
+ * We must explicitly mark this function as used for the compiler,
+ * because otherwise it will be removed from the executable file
+ * during the lto build.
+ */
+API_EXPORT USED int
+on_shutdown_trigger_replace(void *arg, void (*new_handler)(void *),
+			    void (*old_handler)(void *))
+{
+	struct trigger *trigger, *tmp;
+	if (old_handler == NULL) {
+		if (new_handler != NULL) {
+			return on_shutdown_trigger_create(new_handler, arg);
+		} else {
+			/*
+			 * Invalid function params, old_handler or new_handler
+			 * must be set
+			 */
+			errno = EINVAL;
```

Comment:
Don't see much sense in errno setting

Reply:
If you look at the description xtm_api from Mons, then you will see that there is everywhere required to set errno in case of failure. Here it is made by analogy.

Reply:
Sorry for my perfectionism striving, but personally I'd apply this diff:
```
diff --git a/src/on_shutdown.c b/src/on_shutdown.c
index af914ee56..d6cb2aa21 100644
--- a/src/on_shutdown.c
+++ b/src/on_shutdown.c
@@ -80,9 +80,7 @@ box_on_shutdown(void *arg, int (*new_handler)(void *),
 {
        struct on_shutdown_trigger *trigger;
        if (old_handler == NULL) {
-               if (new_handler != NULL) {
-                       return on_shutdown_trigger_create(new_handler, arg);
-               } else {
+               if (new_handler == NULL) {
                        /*
                         * Invalid function params, old_handler or new_handler
                         * must be set
@@ -90,6 +88,7 @@ box_on_shutdown(void *arg, int (*new_handler)(void *),
                        errno = EINVAL;
                        return -1;
                }
+               return on_shutdown_trigger_create(new_handler, arg);
        }
 
        rlist_foreach_entry(trigger, &on_shutdown_trigger_list, link) {
```

Reply:
Can we also use <say> facilities to log error somehow? I'm asking since in two cases we set the same errno, but they are a bit different, so to simplify user's life it would be nice to log error.

Reply:
Add say_error here